File: LeetCode/1415. The k-th Lexicographical String of All Happy Strings of Length n/solution.py
# class Solution:
#     def getHappyString(self, n: int, k: int) -> str:
#         self.n, self.k = n, k
#         self.res, self.kth = "", 1
#         self.dfs("#",0)
#         return self.res

#     def dfs(self,cur,idx):
#         if self.kth > self.k:
#             return
#         if idx == self.n:
#             if self.kth == self.k:
#                 self.res = cur[1:]
#             self.kth += 1
#             return
#         for ch in "abc":
#             if ch != cur[-1]:
#                 self.dfs(cur+ch,idx+1)
#         return


                
class Solution(object):
    def getHappyString(self, n, k):
        # # [!] Brute force, base3 variant
        # for x in range(3 ** n):
        #     # convert x to base 3 sequence
        #     cand = []
        #     for _ in range(n):
        #         if cand and x % 3 == cand[-1]:
        #             break
        #         cand.append(x % 3)
        #         x //= 3
        #     else:
        #         # didn't fail at building the string
        #         k -= 1
        #         if k == 0:
        #             return "".join('abc'[r] for r in cand)[::-1]
        # return ""
        
        # # [!] Brute force, itertools.product variant
        # for cand in itertools.product('abc', repeat=n):
        #     if any(cand[i] == cand[i+1] for i in range(len(cand) - 1)):
        #         continue
        #     k -= 1
        #     if k == 0:
        #         return "".join(cand)
        # return ""
        
        # [!] Backtracking
        ans = [""]
        path = []
        self.k = k
        def search(i):
            if self.k < 0:
                return
            if i == n:
                self.k -= 1
                if self.k == 0:
                    ans.append("".join(path))
                return
            for c in 'abc':
                if not path or path[-1] != c:
                    path.append(c)
                    search(i + 1)
                    path.pop()
        search(0)
        return ans[-1]

File: LeetCode/1415. The k-th Lexicographical String of All Happy Strings of Length n/solution.cpp
// const string alphas = "abc";

// class Solution {
// public:
//     int n, rank;
//     void dfs(int idx, string &path){
//         // cout << idx << " " << path << endl;
//         if (idx == n){
//             rank--;
//             return;
//         }
//         if (rank == 0)
//             return;
//         for (auto ch : alphas){
//             if (!path.empty() && path.back()==ch)
//                 continue;
//             // cout << ch << endl;
//             path.push_back(ch);
//             dfs(idx+1, path);
//             if (rank == 0)
//                 return;
//             path.pop_back();
//         }
//         return;
//     }

//     string getHappyString(int n, int k) {
//         this->n = n;
//         this->rank = k;
//         string path = "";
//         dfs(0,path);
//         return path;
//     }
// };



vector<char> alphas = {'a','b','c'};
vector<vector<char>> mp = {{'b','c'},{'a','c'},{'a','b'}};

class Solution {
public:
    
    string getHappyString(int n, int k) {
        vector<int> pow(n,1);
        for (int i = 1; i < n; i++)
            pow[i] = pow[i-1]*2;
        if (k > 3*pow[n-1])
            return "";
        k--;
        string s;
        int idx = k/pow[n-1];
        s.push_back(alphas[idx]);
        k -= idx*pow[n-1];
        // cout << k << endl;
        for (int i = 1; i < n; i++){
            idx = ((k>>(n-i-1))&1);
            s.push_back(mp[s.back()-'a'][idx]);
        }
        return s;
    }
};